sunpy/io/special/asdf/tags/map/generic_map.py

```diff
@@ -0,0 +1,63 @@
+# -*- coding: utf-8 -*-
+import numpy as np
+
+import astropy.units as u
+from astropy.tests.helper import assert_quantity_allclose
+from asdf.yamlutil import custom_tree_to_tagged_tree
+
+import sunpy.map
+from sunpy.io.special.asdf.types import SunPyType
+
+__all__ = ['GenericMapType']
+
+
+class GenericMapType(SunPyType):
+    name = "map/generic_map"
+    types = ['sunpy.map.GenericMap']
+    requires = ['sunpy']
+    version = "1.0.0"
+
+    @classmethod
+    def from_tree(cls, node, ctx):
+        # Use the factory here to get the correct subclass back
+        out_map = sunpy.map.Map(np.asarray(node['data']), node['meta'])
+        out_map.shift(*node['shift'])
+
+        out_map.mask = node.get('mask')
+        out_map.uncertainty = node.get('uncertainty')
+        out_map._unit = node.get('unit')
+
+        return out_map
+
+    @classmethod
+    def to_tree(cls, smap, ctx):
+        node = {}
+        node['data'] = np.asarray(smap.data)
+        node['meta'] = custom_tree_to_tagged_tree(dict(smap.meta), ctx)
```

Comment:
I'm pretty sure this call to `custom_tree_to_tagged_tree` is redundant given that you call it on `node` at the end of the function and it should apply recursively to all subnodes.